doc/protocol.md

```diff
@@ -829,6 +829,119 @@ or pipe is done.
 Additionally, a "options" control message may be sent in this channel
 to change the "max-size" options.
 
+Payload: fsinfo
+----------------
+
+This channel is used to report information about a given path, and the file or
+a directory which may reside there, including the possibility of updates.  In
+the case of directories, information can also be reported about the (directly)
+contained entries.
+
+Additional "open" command options should be specified with a channel of
+this payload type:
+
+ * `path` (mandatory): a string
+ * `attrs` (mandatory): a list of strings, the list of attributes to be
+   reported for the requested file.  See below for the possibilities.
+ * `fnmatch`: (optional, default ''): if non-empty, this is a [`fnmatch`-style
+   pattern](https://docs.python.org/3/library/fnmatch.html) selecting the directory
+   entries to be reported.  Use `*` to report all files.  Can also be used to
+   implement search (`searc*`) or hide temporary files (`[!.]*`).
+ * `close-on-error` (optional, default: true): if the channel should close
+   (with `problem`) on failure to `open()` the named `path`
+ * `watch` (optional, default: false): if the channel should watch for changes
+ * `follow` (default: true): if the trailing component of `path` is a symbolic
+   link, whether we should follow it.  `no-follow` mode is not supported for
+   `watch` (but is also mostly unnecessary).
+
+A channel is created with an associated absolute pathname ('path').  A path is
+a string that you pass to the `open()` syscall.  The object returned by that
+call is described by the (JSON Document) value of the channel.
+
+Strictly speaking, the path is specified to the channel as a unicode string,
+but paths on the filesystem are binary strings.  This mapping is handled using
+the usual Python convention: filenames are stored in utf8 and encoded and
+decoded with `errors='surrogateescape'`, which allows for handling of filenames
+which are not utf8.
+
+Conceptually, the channel represents the value of a JSON object describing what
+is found at the path, or the reason for an error.  Each message on the channel
+is a JSON Merge Patch (RFC7396) which updates or replaces this object.
+
+The top-level keys in the object are:
+
+ - `info`: the info about the reported path (described below)
+ - `problem`: a "problem code" like `not-found` or `not-directory`
+ - `message`: a user-readable (`strerror`) string describing the error
```

Comment:
:thinking: that might hint devs to show the string to the UI as-is, but (1) that's not translatable, and (2) putting error string comparisons into code is brittle and ugly. Before we commit to that API (which will be difficult to remove), perhaps just give the numeric `errno` or leave it out completely, and we wait until we have a good use case? Do you need this for navigator, or is it sufficient there to just look at `problem` or even just check for success/failure?

Reply:
I thought about this, and almost wrote (English) into the text, but it's not true.  We'll get this in whatever the default language of the user running the bridge is.

In any case, we also report "problem" (and can expand the list of values we have there to match more error codes).

I'm against reporting errno, unless we report it in string form.  These numbers are architecture-specific and quite useless to JSworld.

Reply:
Ah, if they are already translated, then I have no objection. Perhaps mention that in the docs, to discourage devs from trying to match on it?

Reply:
They're actually not already translated, but they theoretically should be.  This is another (rather old) topic about when/if/how we should call `setlocale()`.

Reply:
I checked and it would be possible to report a symbolic errno code like `ENOENT` etc...

Reply:
With the availability of `errno` callers can ignore `message`, or only use it for `console.log` or so, that's good enough for me.